app/shared/design-system/components/primitives/createContentController.js

```diff
@@ -0,0 +1,113 @@
+/**
+ * @description
+ * - detects whether the pointer/keyboard is on any given token as specified by the {contents: []} contract.
+ * - pointermove/keyup in turn dispatch onMouseMove/onCaretMove with the current state if and when within a configured content token.
+ * - returns two methods {update, disconnectedCallback} the former of which dispatches a content update(while preserving the caret position if any)
+ * - read/write methods are called when reading from the passed element or when updating/writing to the same element.
+ * @example
+ * const target = document.querySelector('input');
+ * const {update, disconnectedCallback} = createContentController(target, {
+ *   contents: [
+ *     { name: 'unit', match: (value) => value === 'px' },
+ *     { name: 'number', match: (value) => isNaN(parseFloat(value)) === false },
+ *     { name: 'unknown', match: (value) => true },
+ *   ],
+ *   onMouseMove: ({name, value, position}) => { },
+ *   onCaretMove: ({name, value, position}) => { },
+ *   read: (value) => target.value,
+ *   write: (value) => target.value = value,
+ * });
+ */
+const canvasContext = globalThis.document
+  ?.createElement("canvas")
+  .getContext("2d");
+const measureTextWidth = (value) => canvasContext.measureText(value).width;
+const measureTextPadding = (node, value) => {
+  const {
+    fontSize,
+    fontFamily,
+    fontWeight,
+    textAlign,
+    letterSpacing,
+    paddingLeft,
+    paddingRight,
+  } = getComputedStyle(node);
+  if (letterSpacing !== "normal") canvasContext.letterSpacing = letterSpacing;
+  canvasContext.font = `${fontWeight} ${fontSize} ${fontFamily}`;
+  return textAlign === "left"
+    ? parseFloat(paddingLeft)
+    : node.clientWidth -
+        canvasContext.measureText(value).width -
+        parseFloat(paddingRight);
+};
+const measureTextMetrics = (value, index, offset) => {
+  const DECIMAL_OR_WHITESPACE_OR_NON_WORD = /([-+]?\d*\.?\d+)|([\s\W])/;
+  return value
+    .split(DECIMAL_OR_WHITESPACE_OR_NON_WORD)
+    .filter((value) => value)
+    .map((value) => {
+      const width = measureTextWidth(value);
+      const head = index - offset;
+      const tail = index - (offset + width);
+      const props = { head, tail, value, width, length, offset };
+      length += value.length;
+      offset += width;
+      return props;
+    });
+};
```

Comment:
I wonder if you have considered a more simplistic implementation for this taking into account that we should be using monospaced fonts or rely on OpenType feature `font-variant-numeric: tabular-nums;` for those controls to avoid layout shifts. This would simplify the problem to `const textWidth = numOfChars * CHAR_WIDTH`. For now we could even hard-code the constant and later we could get the value from font metrics or calculated styles once on the app init. WDYT?

Reply:
This would have been so much simpler and avoided bugs, lets add this as a comment inline in code, if this measuring is going to be buggy, we just switch to this simplified solution? Just need to have it here to remember there is a different way, more limited to monospaced fonts, but reliable